### ISSUE TITLE
Response caching middleware doesn't allow injection of IResponseCache

### ISSUE THREAD

**Issue by angrymrt** [aspnet/Home#2957]
The response caching middleware uses a MemoryCache object to cache responses and it is hardcoded in the public constructor to use it. See: https://github.com/aspnet/ResponseCaching/blob/dev/src/Microsoft.AspNetCore.ResponseCaching/ResponseCachingMiddleware.cs. This makes it impossible to provide your own implementation of IResponseCache. I would for example like to create a DistributedResponseCache so I can have better horizontal scaling.

**Comment by tpeczek** [comment]
You can go around this, please take a look here: https://www.tpeczek.com/2018/03/redis-backed-response-caching-in-aspnet.html

**Comment by angrymrt** [comment]
@tpeczek thanks, this is useful as a temporary hack, but I would prefer an official solution for this.

**Comment by tpeczek** [comment]
@angrymrt I went arround, because according to discussion [here](https://github.com/aspnet/ResponseCaching/commit/f125329ed70a5aeaa28d5ecfaf433aed8e318a59#diff-10af08b8eec1eb0744eabb7defa3c966) it has been done on purpose.

**Comment by JunTaoLuo** [comment]
The workaround isn't ideal since it uses reflection but it does get the job done.

**Comment by angrymrt** [comment]
@JunTaoLuo how about a dependency on an IResponseCacheFactory that has a Create(IOptions<ResponseCachingOptions> options) method?

Then just pass the IResponseCacheFactory into the internal constructor and use it to create a cache (either distributed or memory).

**Comment by JunTaoLuo** [comment]
There are certainly many ways to build this functionality. I'm not sure when we'll triage this work item though.

**Comment by yringler** [comment]
@JunTaoLuo , unfortunately that workaround no longer works, because [this commit](https://github.com/dotnet/aspnetcore/commit/cc1f23c5f8afb7d2a00405f19811d2372c4fcec2#diff-bf48ddb7e9ad4ab420a8bf38b9df508d) removes public access to the IResponseCache interface.
Perhaps could that be made public again?

**Comment by slangeder** [comment]
> @JunTaoLuo , unfortunately that workaround no longer works, because [this commit](https://github.com/dotnet/aspnetcore/commit/cc1f23c5f8afb7d2a00405f19811d2372c4fcec2#diff-bf48ddb7e9ad4ab420a8bf38b9df508d) removes public access to the IResponseCache interface.
> Perhaps could that be made public again?

+1 to that. I ended up copying the AspNetCore ResponseCaching code into our repository and adapted the visibility modifiers. Not proud about that, but now we can upgrade to AspNetCore 3.1.
Why can't the ResponseCaching feature make use of the IDistributedCache interface like so many other AspNetCore features?

**Comment by yringler** [comment]
I'm [poking around](https://github.com/yringler/flex-middleware-cache) with maybe making a nuget package which people can use. It uses aspnetcore as a submodule.
The basic idea is to link to all the asnetcore source files through the submodule into a new project, and then add more IResponseCache implementations to that new project.
The only concern is that I end up going down a rabibit hole of linking to more and more source files... But I think it'll be ok.

**Comment by slangeder** [comment]
@yringler Thank you very much for creating that package.
A coworker included it in our project today, and it works perfectly.

**Comment by dustinmoris** [comment]
Possible duplicate or at least closely related to [https://github.com/dotnet/aspnetcore/issues/2617](https://github.com/dotnet/aspnetcore/issues/2617).

Can we priorotise this work again? If I was to send in a PR what are the chances of you guys merging it in for the next release?

**Comment by yringler** [comment]
Yes, a resolution of #2617 would also resolve this. (Unless the resolution is also done with private/internal classes.) It's a particular place where the functionality requested in this issue is needed.

Feel free to take a look at [my repo](https://github.com/yringler/distributed-middleware-cache) mentioned earlier, [available on nuget](https://www.nuget.org/packages/DistributedResponseCachingMiddleware/).

It is a replacement for [Microsoft.AspNetCore.ResponseCaching](https://www.nuget.org/packages/Microsoft.AspNetCore.ResponseCaching/). It's implemented with all the official Microsoft code internally via git submodule and csproj linking, so the amount of code in it which isn't coming from Microsoft is pretty tiny.

@slangeder  is using it, and I have tested it at my company, although we aren't using it in production yet.

If you see something which needs improvement, it's under the MIT license if you want to fork, or you can drop me a PR.

**Comment by JunTaoLuo** [comment]
Extensibility is something we'd want to keep in mind when designing Output Cache: https://github.com/dotnet/aspnetcore/issues/2627.

**Comment by brunolm** [comment]
Any updates on this? What's the current workaround?

**Comment by yringler** [comment]
The only workaround I'm aware of is kinda to fork MS code 🤷 
Even the nice guy I was working with on my proj, after he couldn't use it b/c of his company rules, said that they were going to use the same approach, namely to reference the MS code with CSPROJ links, to access all the internal classes.
There's a workaround [here](https://github.com/yringler/distributed-middleware-cache), which basically uses the MS code, opens it up, and adds some functionality.
I've been meaning to get around to updating the docs and republishing a nuget, but haven't yet. Feel free to take ownership of it!

**Comment by brunolm** [comment]
I found a way to implement a custom filter attribute that can handle server caching. This works for me on dotnet core 5.0

```csharp
namespace Something
{
    [AttributeUsage(AttributeTargets.Method | AttributeTargets.Class, AllowMultiple = false, Inherited = false)]
    public class ActionResultCacheAttribute : ActionFilterAttribute
    {
        private string _cacheKey;

        /// <summary>
        /// The comma separated parameters to vary the caching by.
        /// </summary>
        public string VaryByParam { get; set; }
        public string VaryByHeader { get; set; }
        public string VaryByQueryKeys { get; set; }
        public bool VaryByAuth { get; set; } = false;


        public int Duration { get; set; }
        public int SlidingDuration { get; set; }

        public override async Task OnActionExecutionAsync(ActionExecutingContext context, ActionExecutionDelegate next)
        {
            var _cache = context.HttpContext.RequestServices
                .GetService(typeof(IDistributedCache)) as IDistributedCache;

            this._cacheKey = GetCacheKey(context);

            var resultStr = await _cache.GetStringAsync(this._cacheKey);

            if (resultStr != null)
            {
                var result = Newtonsoft.Json.JsonConvert.DeserializeObject<ObjectResult>(
                    resultStr
                );
                result.Value = result.Value?.ToString();
                context.Result = result;
                return;
            }

            await base.OnActionExecutionAsync(context, next);
        }

        /// <summary>
        /// Occurs when an action has executed.
        /// </summary>
        /// <param name="filterContext">The filter context.</param>
        public async override void OnActionExecuted(ActionExecutedContext filterContext)
        {
            if (filterContext.Exception != null || string.IsNullOrWhiteSpace(this._cacheKey))
            {
                return;
            }

            var _cache = filterContext.HttpContext.RequestServices
                .GetService(typeof(IDistributedCache)) as IDistributedCache;

            var result = Newtonsoft.Json.JsonConvert.SerializeObject(filterContext.Result);

            await _cache.SetStringAsync(this._cacheKey, result, new DistributedCacheEntryOptions
            {
                AbsoluteExpirationRelativeToNow = Duration > 0 ? TimeSpan.FromSeconds(Duration) : null,
                SlidingExpiration = SlidingDuration > 0 ? TimeSpan.FromSeconds(SlidingDuration) : null,
            });
        }

        private string GetCacheKey(ActionExecutingContext context)
        {
            var routeValues = context.RouteData.Values;
            var actionParameters = context.ActionArguments;
            var username = context.HttpContext?.User?.Claims?.FirstOrDefault(o => o.Type == ClaimTypes.Sid)?.Value;

            var sb = new StringBuilder(routeValues["controller"].ToString());
            sb.Append('_').Append(routeValues["action"].ToString());

            if (!string.IsNullOrWhiteSpace(VaryByParam))
            {
                sb.Append(VaryByParam);
            }

            if (!string.IsNullOrWhiteSpace(VaryByHeader))
            {
                sb.Append(VaryByHeader);
            }

            if (!string.IsNullOrWhiteSpace(VaryByQueryKeys))
            {
                sb.Append(VaryByQueryKeys);
            }

            if (VaryByAuth)
            {
                sb.Append($"U-{username}-");
            }

            return sb.ToString();
        }
    }
}
```

**Comment by Ognen67** [comment]
> The only workaround I'm aware of is kinda to fork MS code 🤷 Even the nice guy I was working with on my proj, after he couldn't use it b/c of his company rules, said that they were going to use the same approach, namely to reference the MS code with CSPROJ links, to access all the internal classes. There's a workaround [here](https://github.com/yringler/distributed-middleware-cache), which basically uses the MS code, opens it up, and adds some functionality. I've been meaning to get around to updating the docs and republishing a nuget, but haven't yet. Feel free to take ownership of it!

Does this still work now with the Microsoft.AspNetCore.ResponseCaching package Version: 2.2.0 ?

**Comment by Ognen67** [comment]
> > Does this still work now with the Microsoft.AspNetCore.ResponseCaching package Version: 2.2.0 ?
> 
> That NuGet package is deprecated and doesn't receive updates.

Okay, thanks for the prompt response.

**Comment by Ognen67** [comment]
Is there another way to achieve the same results as of now?
